packages/@aws-cdk/aws-stepfunctions-tasks/lib/emrcontainers/start-job-run.ts

```diff
@@ -0,0 +1,166 @@
+import * as iam from '@aws-cdk/aws-iam';
+import * as sfn from '@aws-cdk/aws-stepfunctions';
+import * as cdk from '@aws-cdk/core';
+import { Construct } from 'constructs';
+import { integrationResourceArn, validatePatternSupported } from '../private/task-utils';
+import { JobDriver, ConfigurationOverrides, ReleaseLabel } from './base-types';
+
+/**
+ * The props for a EMR Containers StartJobRun Task.
+ */
+export interface EmrContainersStartJobRunProps extends sfn.TaskStateBaseProps {
+
+  /**
+   * The virtual cluster ID for which the job run request is submitted.
+   */
+  readonly virtualClusterId: sfn.TaskInput;
+
+  /**
+   * The name of the job run.
+   *
+   * @default - No job name
+   */
+  readonly jobName?: string;
+
+  /**
+   * The execution role ARN for the job run.
+   *
+   * @default - No execution arn
+   */
+  readonly executionRoleArn: string;
+
+  /**
+   * The Amazon EMR release version to use for the job run.
+   */
+  readonly releaseLabel: ReleaseLabel;
+
+  /**
+   * The job driver for the job run.
+   *
+   * @see https://docs.aws.amazon.com/emr-on-eks/latest/APIReference/API_JobDriver.html
+   *
+   * @default - No job driver
+   */
+  readonly jobDriver: JobDriver;
+
+  /**
+   * A configuration specification to be used to override existing configurations.
+   *
+   * @see https://docs.aws.amazon.com/emr-on-eks/latest/APIReference/API_ConfigurationOverrides.html
+   *
+   * @default - No configuration override
+   */
+  readonly configurationOverrides?: ConfigurationOverrides;
+
+  /**
+   * The tags assigned to job runs.
+   *
+   * @default - None
+   */
+  readonly tags?: { [key: string]: string };
+}
+
+/**
+ * Starts a job run. A job run is a unit of work, such as a Spark jar, PySpark script, or SparkSQL query,
+ * that you submit to Amazon EMR on EKS.
+ *
+ * @see https://docs.amazonaws.cn/en_us/step-functions/latest/dg/connect-emr-eks.html
+ */
+export class EmrContainersStartJobRun extends sfn.TaskStateBase {
```

Comment:
```ts
interface Monitoring: {
  logging?: boolean;
  logGroup?: logs.ILogGroup;
  logStream?: logs.ILogStream; // added log stream
  logBucket?: s3.IBucket;
  persistentAppUI?: boolean;
}
```

Can you tell me the benefits behind automatically creating the loggroup/stream/ bucket? I don't really understand the user benefits behind this. I'm also having trouble implementing it because I believe instantiating them in the constructor would be cost ineffective

Something like
```ts

protected logGroup: logs.ILogGroup;
  protected logStream: logs.ILogStream;
  protected s3Bucket: s3.IBucket;

  private readonly integrationPattern: sfn.IntegrationPattern;

  constructor(scope: Construct, id: string, private readonly props: EmrContainersStartJobRunProps) {
    super(scope, id, props);
    this.integrationPattern = props.integrationPattern ?? sfn.IntegrationPattern.RUN_JOB;

    validatePatternSupported(this.integrationPattern, EmrContainersStartJobRun.SUPPORTED_INTEGRATION_PATTERNS);

    this.logGroup = new logs.LogGroup(scope, 'EMR Containers Default Cloudwatch Log Group');
    this.logStream = new logs.LogStream(this, 'EMR Containers Default Cloudwatch Log Stream Prefix', {
      logGroup: this.logGroup,
    });
    this.s3Bucket = new s3.Bucket(scope, 'EMR Containers Default S3 bucket', {
      bucketName: 'DefaultEmrContainersBucket',
    });

    this.taskPolicies = this.createPolicyStatements();
  }
```

and then rendering them in SFN is tricky too.

```ts

Reply:
https://docs.aws.amazon.com/cdk/api/latest/typescript/api/aws-s3/bucketprops.html#aws_s3_BucketProps_bucketName

Found the documentation that shows cloudformation automatically generates this

Reply:
>```logStream?: logs.ILogStream; // added log stream```

The API only takes a stream name prefix, not an entire log stream. That's to say, even if we ask for and receive a log stream here and use the log stream name as our prefix, the emr-containers service will create a new log stream. We *could* add a `logStreamPrefix: string` property here but I find that overkill, personally. Happy to discuss.

> If we default this there will be a conflict in names, so are there any potential solutions to this?

As you mentioned, there is no need to provide a name for the bucket, nor for any auto-generated resource we make for the user.

> Additionally, the execution role cannot be created automatically as there's no way for EMR on EKS to do so via the CDK, because they must follow a set of steps in setting up that can't be done in the CDK.

I'm not sure if this is entirely true. I've taken a look at the steps required, and it seems like nearly everything can be done using some combination of kubectl and various AWS APIs. These steps can all be run from one or more [custom resources](https://docs.aws.amazon.com/cdk/api/latest/docs/custom-resources-readme.html), allowing for a zero-touch setup.

The exception is possibly [enabling IAM roles for service accounts](https://docs.aws.amazon.com/emr/latest/EMR-on-EKS-DevelopmentGuide/setting-up-enable-IAM.html), where [creating an OIDC identity provider](https://docs.aws.amazon.com/IAM/latest/UserGuide/id_roles_providers_create_oidc.html#manage-oidc-provider-api) via the AWS IAM API may [require some tool to investigate a server's certificate](https://docs.aws.amazon.com/IAM/latest/UserGuide/id_roles_providers_create_oidc_verify-thumbprint.html), usually OpenSSL but possibly there are other options. One alternative would be to use [eksctl](https://github.com/weaveworks/eksctl) (the supported CLI for EMR containers) which could simplify some of the other setup as well, but I'll have to consult the team on if that is a lower risk/operational burden. Some more research is required on your end into what can be done automatically without the user taking any manual actions.

> ```monitoringConfiguration?: MonitoringConfiguration;```

Fairly minor, but I'd rather this be kept as just `monitoring: Monitoring`. I think "configuration" as a word is overused and doesn't have much meaning here. By the nature of the CDK, we're configuring a good many things. "Application configuration" makes sense in this context because the application is being run by EMR and we are *only* configuring it here, not specifying in its entirety. On the other hand, we are fully specifying the monitoring for the job run and so can drop the "config" suffix. What do you think? There's an argument for consistency...

Reply:
> The API only takes a stream name prefix, not an entire log stream. That's to say, even if we ask for and receive a log   stream here and use the log stream name as our prefix, the emr-containers service will create a new log stream. We could add a `logStreamPrefix: string` property here but I find that overkill, personally. Happy to discuss.

To open up a discussion on this topic, we wouldn't want our advanced/beginner users to be confused by the lack of a logStreamPrefix property, for example, EMR on EKS's [tutorial ](https://docs.aws.amazon.com/emr/latest/EMR-on-EKS-DevelopmentGuide/getting-started.html)explicitly asks for a logStreamPrefix. Additionally, in my meetings with my team we've discussed that it's imperative that nearly/all of the features that we have available for our API be present in CDK implementation. I understand that it feels overkill, but we would want users to be able to do the same thing on the CDK as elsewhere.

> I'm not sure if this is entirely true. I've taken a look at the steps required, and it seems like nearly everything can be done using some combination of kubectl and various AWS APIs. These steps can all be run from one or more [custom resources](https://docs.aws.amazon.com/cdk/api/latest/docs/custom-resources-readme.html), allowing for a zero-touch setup.
> 
> The exception is possibly [enabling IAM roles for service accounts](https://docs.aws.amazon.com/emr/latest/EMR-on-EKS-DevelopmentGuide/setting-up-enable-IAM.html), where [creating an OIDC identity provider](https://docs.aws.amazon.com/IAM/latest/UserGuide/id_roles_providers_create_oidc.html#manage-oidc-provider-api) via the AWS IAM API may [require some tool to investigate a server's certificate](https://docs.aws.amazon.com/IAM/latest/UserGuide/id_roles_providers_create_oidc_verify-thumbprint.html), usually OpenSSL but possibly there are other options. One alternative would be to use [eksctl](https://github.com/weaveworks/eksctl) (the supported CLI for EMR containers) which could simplify some of the other setup as well, but I'll have to consult the team on if that is a lower risk/operational burden. Some more research is required on your end into what can be done automatically without the user taking any manual actions.
> 

I agree with you, because I spent most of yesterday also investigating the same alternative solution that would help automate the setting up process, and brought it up during my mentor and standup meeting. However, the general consensus I've received from both my team and mentor is that they're looking to simplify the setting up process themselves and that it's out of scope for the CDK implementation. Thank you for bringing this to our attention though, I don't believe the team knew about this. This will be a valuable data point in the future for the simplification of the setting up phase. For now, it's expected that user's finish the setting up steps for EKS at least. 

On the other hand, the team and  I am happy with the idea of an automatically generated IAM role that we can provide with correct permissions / and a call to update-role-trust-policy. I can move forward on this idea.

Reply:
> Log streams

Okay, let's go with `logStreamPrefix?: string`

> Automated cluster setup

Alright, let's operate under the assumption that the cluster has been configured properly. We will need to make sure we clearly document the steps required out of band. Let's go ahead with just the execution role creation + trust policy update.

I think we are aligned on the API, do you have any outstanding questions? Otherwise, let's get started on implementation.

Reply:
Great will do. I was implementing everything as we discussed it, so I just need to get the unit tests for start-job-run and the execution role creation + trust policy implemented.